libgalois/include/katana/PropertyGraph.h

```diff
@@ -842,6 +852,513 @@ CreateSymmetricGraph(PropertyGraph* pg);
+  EdgeTypeIndex& operator=(EdgeTypeIndex&&) = default;
+
+  EdgeTypeIndex(const EdgeTypeIndex&) = delete;
+  EdgeTypeIndex& operator=(const EdgeTypeIndex&) = delete;
+
+  static EdgeTypeIndex Make(const PropertyGraph* pg) noexcept;
+
+  EdgeTypeID GetType(uint32_t index) const noexcept {
+    KATANA_LOG_ASSERT(index < num_unique_edge_types_);
+    KATANA_LOG_ASSERT(size_t(index) < edge_index_to_type_map_.size());
+    return edge_index_to_type_map_[index];
+  }
+
+  uint32_t GetIndex(const EdgeTypeID& edge_type) const noexcept {
+    KATANA_LOG_ASSERT(edge_type_to_index_map_.count(edge_type) > 0);
+    // return edge_type_to_index_map_[edge_type];
+    return edge_type_to_index_map_.find(edge_type)->second;
+  }
+
+  uint32_t num_unique_types() const noexcept { return num_unique_edge_types_; }
+
+  /// @param edge_type: edge_type to check
+  /// @returns true iff there exists some edge in the graph with that edge_type
+  bool has_edge_type_id(const EdgeTypeID& edge_type) const noexcept {
+    return (
+        edge_type_to_index_map_.find(edge_type) !=
+        edge_type_to_index_map_.cend());
+  }
+
+  /// Wrapper to get the distinct edge types in the graph.
+  ///
+  /// @returns Range of the distinct edge types
+  EdgeTypeIDRange distinct_edge_type_ids() const noexcept {
+    return EdgeTypeIDRange{
+        edge_index_to_type_map_.cbegin(), edge_index_to_type_map_.cend()};
+  }
+
+private:
+  EdgeTypeIndex(
+      EdgeTypeIDToIndexMap&& edge_type_to_index,
+      EdgeIndexToTypeIDMap&& edge_index_to_type,
+      uint32_t num_edge_types) noexcept
+      : edge_type_to_index_map_(std::move(edge_type_to_index)),
+        edge_index_to_type_map_(std::move(edge_index_to_type)),
+        num_unique_edge_types_(num_edge_types) {
+    KATANA_LOG_DEBUG_ASSERT(
+        num_edge_types == edge_index_to_type_map_.size() &&
+        num_edge_types == edge_type_to_index_map_.size());
+  }
+
+  EdgeTypeIDToIndexMap edge_type_to_index_map_;
+  EdgeIndexToTypeIDMap edge_index_to_type_map_;
+  uint32_t num_unique_edge_types_ = 0u;
+};
+
+// TODO(amber): make OrigEdgeIDMap optional via template argument
+class KATANA_EXPORT EdgeShuffleTopology : public GraphTopology {
+  using Base = GraphTopology;
+  using OrigEdgeIDMap = katana::NUMAArray<Edge>;
+  using EdgeTypeID = PropertyGraph::TypeSetID;
+
+public:
+  EdgeShuffleTopology() = default;
+  EdgeShuffleTopology(EdgeShuffleTopology&&) = default;
+  EdgeShuffleTopology& operator=(EdgeShuffleTopology&&) = default;
+
+  EdgeShuffleTopology(const EdgeShuffleTopology&) = delete;
+  EdgeShuffleTopology& operator=(const EdgeShuffleTopology&) = delete;
+
+  static EdgeShuffleTopology MakeTransposeCopy(const PropertyGraph* pg);
+  static EdgeShuffleTopology MakeOriginalCopy(const PropertyGraph* pg);
+
+  auto original_edge_id(Edge eid) const noexcept {
+    KATANA_LOG_DEBUG_ASSERT(eid < orig_edge_ids_.size());
+    return orig_edge_ids_[eid];
+  }
+
+  void SortEdgesByTypeThenDest() noexcept;
```

Comment:
nit: I'd prefer `SortedEdgeTopology` or something of that sort instead of shuffle.

Reply:
My current thinking is to use shuffle as umbrella term for both sorting. We could argue that things like transpose are a sorting afterall. So let me think about it and may be I'll rename it to sorted in the next PR.

Reply:
I didn't realize that `Shuffle` includes transpose. Transpose has iteration over in-edge but sorting doesn't, so how/why are they under the same class?

Reply:
Think of (going forward as well, for next PR) `Shuffle` as a building block for optional topologies. Transpose is one way to shuffle (or sort) edges. 
Secondly, the concept of `in-edges` only applies to bi-directional topologies. Transpose as a stand-alone doesn't support `in_edges()`, it just supports `edges()`, because it's a building block to something else (e.g., a bidirectional topology). Say I take GraphTopology and transpose it somehow, it's `edges(src)` method will start return `in_edges()`.

Reply:
Ok. That makes sense. Please document it in the code in a subsequent PR.